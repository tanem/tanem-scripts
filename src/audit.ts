import execa from 'execa'
import glob from 'glob'
import path from 'path'

const exec = (file: string, args?: readonly string[]) => {
  execa.sync(file, args, { stdio: 'inherit' })
}

export const audit = ({ fix }: { fix?: boolean } = {}) => {
  glob
    .sync('**/package.json', {
      ignore: '**/node_modules/**',
      absolute: true
    })
    .map(path.dirname)
    .forEach(dir => {
      process.chdir(dir)
      exec('npm', ['i'])
      exec('npm', ['audit', ...(fix ? ['fix'] : [])])
    })
}
