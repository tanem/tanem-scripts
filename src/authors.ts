import execa from 'execa';

interface Options {
  isNumbered?: boolean;
}

const clean = (result: string) =>
  result
    .split('\n')
    .map(line => line.replace(/ *\d+\t/g, ''))
    .join('\n');

const authors = async ({ isNumbered }: Options = {}) => {
  const { stdout } = await execa('git', [
    'shortlog',
    'HEAD',
    `-se${isNumbered ? 'n' : ''}`
  ]);
  return clean(stdout);
};

authors.sync = ({ isNumbered }: Options = {}) => {
  const { stdout } = execa.sync('git', [
    'shortlog',
    'HEAD',
    `-se${isNumbered ? 'n' : ''}`
  ]);
  return clean(stdout);
};

export default authors;
