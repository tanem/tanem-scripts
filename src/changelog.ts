import Octokit from '@octokit/rest';
import { compareAsc, format, isBefore } from 'date-fns';
import gitRemoteOriginUrl from 'git-remote-origin-url';
import parseGithubUrl from 'parse-github-url';

const getRepoInfo = async () => {
  const url = await gitRemoteOriginUrl();
  const parsed = parseGithubUrl(url);

  /* istanbul ignore else */
  if (parsed && parsed.owner && parsed.name) {
    return {
      owner: parsed.owner,
      repo: parsed.name
    };
  }

  throw new Error('Unable to parse GitHub url');
};

const changelog = async ({
  futureRelease,
  owner,
  repo
}: {
  futureRelease?: string;
  owner?: string;
  repo?: string;
} = {}) => {
  if (!owner || !repo) {
    const repoInfo = await getRepoInfo();

    /* istanbul ignore else */
    if (!owner) {
      owner = repoInfo.owner;
    }

    /* istanbul ignore else */
    if (!repo) {
      repo = repoInfo.repo;
    }
  }

  const octokit = new Octokit({
    auth: process.env.CHANGELOG_GITHUB_TOKEN
  });

  const baseEndpointOptions = {
    owner,
    per_page: 100, // eslint-disable-line @typescript-eslint/camelcase
    repo
  };

  const [rawPulls, rawTags, rawCommits]: [
    Octokit.PullsListResponseItem[],
    Octokit.ReposListTagsResponseItem[],
    Octokit.ReposListCommitsResponseItem[]
  ] = await Promise.all([
    octokit.paginate(
      octokit.pulls.list.endpoint.merge({
        ...baseEndpointOptions,
        state: 'closed'
      })
    ),
    octokit.paginate(
      octokit.repos.listTags.endpoint.merge(baseEndpointOptions)
    ),
    octokit.paginate(
      octokit.repos.listCommits.endpoint.merge(baseEndpointOptions)
    )
  ]);

  const cleanedPulls = rawPulls
    .filter(pull => Boolean(pull.merged_at))
    .map(pull => ({
      mergedAt: pull.merged_at,
      number: pull.number,
      title: pull.title,
      userHtmlUrl: pull.user.html_url,
      userLogin: pull.user.login
    }));
  cleanedPulls.sort((a, b) => compareAsc(a.mergedAt, b.mergedAt));

  const pendingCleanedTags = [];

  const getCleanTagData = async (
    tag: Octokit.ReposListTagsResponseItem,
    owner: string,
    repo: string
  ): Promise<{
    date: string;
    name: string;
    pulls: typeof cleanedPulls;
  }> => {
    const existingTagCommit = rawCommits.find(
      commit => commit.sha === tag.commit.sha
    );

    if (existingTagCommit) {
      return {
        date: existingTagCommit.commit.committer.date,
        name: tag.name,
        pulls: []
      };
    }

    const { data: latestTagCommit } = await octokit.git.getCommit({
      commit_sha: tag.commit.sha, // eslint-disable-line @typescript-eslint/camelcase
      owner,
      repo
    });

    return {
      date: latestTagCommit.committer.date,
      name: tag.name,
      pulls: []
    };
  };

  for (const tag of rawTags) {
    pendingCleanedTags.push(getCleanTagData(tag, owner, repo));
  }

  const cleanedTags = await Promise.all(pendingCleanedTags);
  if (futureRelease) {
    cleanedTags.unshift({
      date: new Date().toISOString(),
      name: futureRelease,
      pulls: []
    });
  }
  cleanedTags.sort((a, b) => compareAsc(a.date, b.date));

  cleanedPulls.map(p => {
    const tag = cleanedTags.find(t => isBefore(p.mergedAt, t.date));
    /* istanbul ignore else */
    if (tag) {
      tag.pulls.unshift(p);
    }
  });

  cleanedTags.reverse();

  return (
    '# Changelog\n' +
    cleanedTags
      .map((tag, index, array) => {
        let result = `\n## [${
          tag.name
        }](https://github.com/${owner}/${repo}/tree/${tag.name}) (${format(
          tag.date,
          'YYYY-MM-DD'
        )})\n`;

        if (index + 1 !== array.length) {
          result += `[Full Changelog](https://github.com/${owner}/${repo}/compare/${array[index + 1].name}...${tag.name})\n`;
        }

        tag.pulls.map((pull, index: number) => {
          if (index === 0) {
            result += `\n**Merged pull requests:**\n\n`;
          }
          result += `- ${pull.title} [\#${pull.number}](https://github.com/${owner}/${repo}/pull/${pull.number}) ([${pull.userLogin}](${pull.userHtmlUrl}))\n`;
        });

        return result;
      })
      .join('')
  );
};

export default changelog;
