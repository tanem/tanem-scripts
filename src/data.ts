import Octokit from '@octokit/rest';
import { compareAsc } from 'date-fns';
import gitRemoteOriginUrl from 'git-remote-origin-url';
import parseGithubUrl from 'parse-github-url';

export interface Data {
  commits: Octokit.ReposListCommitsResponseItem[];
  owner: string;
  pulls: Octokit.PullsListResponseItem[];
  repo: string;
  tags: {
    date: string;
    name: string;
  }[];
}

const getRepoInfo = async () => {
  const url = await gitRemoteOriginUrl();
  const parsed = parseGithubUrl(url);

  /* istanbul ignore else */
  if (parsed && parsed.owner && parsed.name) {
    return {
      owner: parsed.owner,
      repo: parsed.name
    };
  }

  throw new Error('Unable to parse GitHub url');
};

export const cache = new Map<'data', Data>();

export const get = async (): Promise<Data> => {
  let data = cache.get('data');

  if (data) {
    return data;
  }

  const { owner, repo } = await getRepoInfo();

  const octokit = new Octokit({
    auth: process.env.CHANGELOG_GITHUB_TOKEN
  });

  const baseEndpointOptions = {
    owner,
    per_page: 100, // eslint-disable-line @typescript-eslint/camelcase
    repo
  };

  const [rawPulls, rawTags, commits]: [
    Octokit.PullsListResponseItem[],
    Octokit.ReposListTagsResponseItem[],
    Octokit.ReposListCommitsResponseItem[]
  ] = await Promise.all([
    octokit.paginate(
      octokit.pulls.list.endpoint.merge({
        ...baseEndpointOptions,
        state: 'closed'
      })
    ),
    octokit.paginate(
      octokit.repos.listTags.endpoint.merge(baseEndpointOptions)
    ),
    octokit.paginate(
      octokit.repos.listCommits.endpoint.merge(baseEndpointOptions)
    )
  ]);

  const pulls = rawPulls.filter(pull => Boolean(pull.merged_at));

  pulls.sort((a, b) =>
    compareAsc(new Date(a.merged_at), new Date(b.merged_at))
  );

  const tags = await Promise.all(
    rawTags.map(async tag => {
      const commit = commits.find(commit => commit.sha === tag.commit.sha);

      if (commit) {
        return {
          date: commit.commit.committer.date,
          name: tag.name
        };
      }

      const { data: tagCommit } = await octokit.git.getCommit({
        commit_sha: tag.commit.sha, // eslint-disable-line @typescript-eslint/camelcase
        owner,
        repo
      });

      return {
        date: tagCommit.committer.date,
        name: tag.name
      };
    })
  );

  tags.sort((a, b) => compareAsc(new Date(a.date), new Date(b.date)));

  data = {
    commits,
    owner,
    pulls,
    repo,
    tags
  };

  cache.set('data', data);

  return data;
};
