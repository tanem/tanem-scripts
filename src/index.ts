export { default as authors } from './authors';
export { default as changelog } from './changelog';
