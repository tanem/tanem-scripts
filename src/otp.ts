import promptPromise from 'prompt-promise';

export enum Messages {
  PROMPT = 'NPM 2-factor auth code: ',
  HELP = 'Two-factor auth is required to publish.',
}

// Hat-tip: https://github.com/facebook/react/blob/master/scripts/release/publish-commands/prompt-for-otp.js.
// TODO (Tane): This'll be nice to do at some point: https://github.com/facebook/react/blob/master/scripts/release/theme.js.
export const prompt = async () => {
  while (true) {
    const otp = await promptPromise(Messages.PROMPT);
    promptPromise.done();

    if (otp) {
      return otp;
    } else {
      console.log(Messages.HELP);
    }
  }
};
