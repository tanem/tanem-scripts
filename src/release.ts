import { isAfter } from 'date-fns';
import execa from 'execa';
import { get as getData } from './data';

const release = async () => {
  const { pulls, tags } = await getData();

  const latestTag = tags.pop();

  const pullsToRelease = latestTag
    ? pulls.filter(pull =>
        isAfter(new Date(pull.merged_at), new Date(latestTag.date))
      )
    : pulls;

  const labelsToRelease = [
    ...new Set(
      pullsToRelease
        .filter(pull => pull.labels.length)
        .map(pull => {
          const firstLabel = pull.labels.pop();
          if (firstLabel) {
            // TODO: Throw here?
            return firstLabel.name;
          }
        })
    )
  ];

  const releaseType = labelsToRelease.includes('breaking')
    ? 'major'
    : labelsToRelease.includes('enhancement')
    ? 'minor'
    : 'patch';

  const result = execa('npm', ['version', releaseType, '-m', 'Release v%s']);

  result.stdout &&
    result.stdout.pipe(
      process.stdout,
      { end: false }
    );

  result.stderr &&
    result.stderr.pipe(
      process.stderr,
      { end: false }
    );
};

export default release;
