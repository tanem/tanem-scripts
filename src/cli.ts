#!/usr/bin/env node

import program from 'commander';
import authors from './authors';
import changelog from './changelog';
import release from './release';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const { description, version } = require('../package.json');

program
  .version(version)
  .name('tanem-scripts')
  .description(description);

program.on('command:*', function() {
  console.error(
    'Invalid command: %s\nSee --help for a list of available commands.',
    program.args.join(' ')
  );
  process.exit(1);
});

program
  .command('authors')
  .description(
    'generates a list of authors in a format suitable for inclusion in an AUTHORS file'
  )
  .option('-n, --numbered', 'sort by number of commits per author')
  .action(async cmd => {
    try {
      const result = await authors({ isNumbered: cmd.numbered });
      process.stdout.write(result);
    } catch (error) {
      console.error(error);
      process.exit(1);
    }
  })
  .on('--help', () => {
    console.log(`
Examples:
  $ authors
  $ authors -n`);
  });

program
  .command('changelog')
  .description('generates a changelog using GitHub tags and pull requests')
  .option(
    '-f, --future-release <tag>',
    'tag to use for PRs merged since the last published tag'
  )
  .option('-o, --owner <owner>', 'repo owner')
  .option('-r, --repo <repo>', 'repo name')
  .action(async cmd => {
    try {
      const result = await changelog({
        futureRelease: cmd.futureRelease,
        owner: cmd.owner,
        repo: cmd.repo
      });
      process.stdout.write(result);
    } catch (error) {
      console.error(error);
      process.exit(1);
    }
  })
  .on('--help', () => {
    console.log(`
  Examples:
    $ changelog -f v1.0.0
    $ changelog -o tanem -r react-svg
    $ changelog -f v2.0.0 -o tanem -r react-svg
  `);
  });

program
  .command('release')
  .description('publishes a package to npm')
  .option('-o, --owner <owner>', 'repo owner')
  .option('-r, --repo <repo>', 'repo name')
  .action(async cmd => {
    try {
      const result = await release({
        owner: cmd.owner,
        repo: cmd.repo
      });
      process.stdout.write(result);
    } catch (error) {
      console.error(error);
      process.exit(1);
    }
  })
  .on('--help', () => {
    console.log(`
  Examples:
    $ release -o tanem -r react-svg
  `);
  });

program.parse(process.argv);
