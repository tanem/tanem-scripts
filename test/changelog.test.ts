/* eslint-disable @typescript-eslint/camelcase */

import { subDays, subHours } from 'date-fns';
import fs from 'fs';
import lolex, { InstalledClock } from 'lolex';
import nock from 'nock';
import path from 'path';
import { changelog } from '../src';

nock.disableNetConnect();

const originalChangelogGithubToken = process.env.CHANGELOG_GITHUB_TOKEN;
const originalGitDir = process.env.GIT_DIR;
const owner = 'owner';
const repo = 'repo';
const userHtmlUrl = 'https://github.com/user';
const userLogin = 'user';
const tmpDirPath = path.join(__dirname, 'tmp-changelog');
const gitConfigPath = path.join(tmpDirPath, 'config');

let clock: InstalledClock;
let now: string;

beforeEach(() => {
  fs.mkdirSync(tmpDirPath);
  fs.writeFileSync(
    gitConfigPath,
    `[remote "origin"]\n  url = git@github.com:${owner}/${repo}.git`,
    'utf-8'
  );
  process.env.CHANGELOG_GITHUB_TOKEN = 'token';
  process.env.GIT_DIR = path.join(__dirname, 'tmp-changelog');
  clock = lolex.install({ now: new Date(2000, 1, 1) });
  now = new Date().toISOString();
});

afterEach(() => {
  fs.unlinkSync(gitConfigPath);
  fs.rmdirSync(tmpDirPath);
  process.env.CHANGELOG_GITHUB_TOKEN = originalChangelogGithubToken;
  process.env.GIT_DIR = originalGitDir;
  clock.uninstall();
});

test('handles GitHub API pagination', async () => {
  nock('https://api.github.com')
    .get('/repos/owner/repo/pulls')
    .query({ per_page: 100, state: 'closed' })
    .reply(
      200,
      [
        {
          id: 2,
          merged_at: subHours(now, 1),
          number: 2,
          title: 'Pull request 2',
          user: {
            html_url: userHtmlUrl,
            login: userLogin
          }
        }
      ],
      {
        Link:
          '<https://api.github.com/repos/owner/repo/pulls?page=2&per_page=100&state=closed>; rel="next"'
      }
    )
    .get('/repos/owner/repo/pulls')
    .query({ page: 2, per_page: 100, state: 'closed' })
    .reply(200, [
      {
        id: 1,
        merged_at: subDays(subHours(now, 1), 1),
        number: 1,
        title: 'Pull request 1',
        user: {
          html_url: userHtmlUrl,
          login: userLogin
        }
      }
    ]);

  nock('https://api.github.com')
    .get('/repos/owner/repo/tags')
    .query({ per_page: 100 })
    .reply(200, [{ name: 'v1.0.0', commit: { sha: '1' } }], {
      Link:
        '<https://api.github.com/repos/owner/repo/tags?page=2&per_page=100>; rel="next"'
    })
    .get('/repos/owner/repo/tags')
    .query({ page: 2, per_page: 100 })
    .reply(200, [{ name: 'v2.0.0', commit: { sha: '2' } }]);

  nock('https://api.github.com')
    .get('/repos/owner/repo/commits')
    .query({ per_page: 100 })
    .reply(
      200,
      [
        {
          sha: '2',
          commit: {
            committer: {
              date: now
            }
          }
        }
      ],
      {
        Link:
          '<https://api.github.com/repos/owner/repo/commits?page=2&per_page=100>; rel="next"'
      }
    )
    .get('/repos/owner/repo/commits')
    .query({ page: 2, per_page: 100 })
    .reply(200, [
      {
        sha: '1',
        commit: {
          committer: {
            date: subDays(now, 1)
          }
        }
      }
    ]);

  const result = await changelog({
    owner,
    repo
  });

  expect(result).toMatchSnapshot();
});

test('only uses merged PRs', async () => {
  nock('https://api.github.com')
    .get('/repos/owner/repo/pulls')
    .query({ per_page: 100, state: 'closed' })
    .reply(200, [
      {
        id: 2,
        merged_at: subHours(now, 2),
        number: 2,
        title: 'Pull request 2',
        user: {
          html_url: userHtmlUrl,
          login: userLogin
        }
      },
      {
        id: 1,
        merged_at: subDays(subHours(now, 2), 1),
        number: 1,
        title: 'Pull request 1',
        user: {
          html_url: userHtmlUrl,
          login: userLogin
        }
      },
      {
        id: 3,
        merged_at: null,
        number: 3,
        title: 'Pull request 3',
        user: {
          html_url: userHtmlUrl,
          login: userLogin
        }
      }
    ]);

  nock('https://api.github.com')
    .get('/repos/owner/repo/tags')
    .query({ per_page: 100 })
    .reply(200, [
      { name: 'v1.0.0', commit: { sha: '1' } },
      { name: 'v2.0.0', commit: { sha: '2' } }
    ]);

  nock('https://api.github.com')
    .get('/repos/owner/repo/commits')
    .query({ per_page: 100 })
    .reply(200, [
      {
        sha: '2',
        commit: {
          committer: {
            date: subHours(now, 1)
          }
        }
      },
      {
        sha: '1',
        commit: {
          committer: {
            date: subDays(subHours(now, 1), 1)
          }
        }
      }
    ]);

  const result = await changelog({
    owner,
    repo
  });

  expect(result).toMatchSnapshot();
});

test('refetches missing tag commits', async () => {
  nock('https://api.github.com')
    .get('/repos/owner/repo/pulls')
    .query({ per_page: 100, state: 'closed' })
    .reply(200, [
      {
        id: 1,
        merged_at: subHours(now, 1),
        number: 1,
        title: 'Pull request 1',
        user: {
          html_url: userHtmlUrl,
          login: userLogin
        }
      },
      {
        id: 2,
        merged_at: subDays(subHours(now, 1), 1),
        number: 2,
        title: 'Pull request 2',
        user: {
          html_url: userHtmlUrl,
          login: userLogin
        }
      }
    ]);

  nock('https://api.github.com')
    .get('/repos/owner/repo/tags')
    .query({ per_page: 100 })
    .reply(200, [
      { name: 'v1.0.0', commit: { sha: '1' } },
      { name: 'v2.0.0', commit: { sha: '2' } }
    ]);

  nock('https://api.github.com')
    .get('/repos/owner/repo/commits')
    .query({ per_page: 100 })
    .reply(200, [
      {
        sha: '2',
        commit: {
          committer: {
            date: now
          }
        }
      }
    ]);

  nock('https://api.github.com')
    .get('/repos/owner/repo/git/commits/1')
    .reply(200, {
      sha: '1',
      committer: {
        date: subDays(now, 1)
      }
    });

  const result = await changelog({
    owner,
    repo
  });

  expect(result).toMatchSnapshot();
});

test('can create a future release', async () => {
  nock('https://api.github.com')
    .get('/repos/owner/repo/pulls')
    .query({ per_page: 100, state: 'closed' })
    .reply(200, [
      {
        id: 2,
        merged_at: subHours(now, 1),
        number: 2,
        title: 'Pull request 2',
        user: {
          html_url: userHtmlUrl,
          login: userLogin
        }
      },
      {
        id: 1,
        merged_at: subHours(now, 3),
        number: 1,
        title: 'Pull request 1',
        user: {
          html_url: userHtmlUrl,
          login: userLogin
        }
      }
    ]);

  nock('https://api.github.com')
    .get('/repos/owner/repo/tags')
    .query({ per_page: 100 })
    .reply(200, [{ name: 'v1.0.0', commit: { sha: '1' } }]);

  nock('https://api.github.com')
    .get('/repos/owner/repo/commits')
    .query({ per_page: 100 })
    .reply(200, [
      {
        sha: '1',
        commit: {
          committer: {
            date: subHours(now, 2)
          }
        }
      }
    ]);

  const result = await changelog({
    futureRelease: 'v2.0.0',
    owner,
    repo
  });

  expect(result).toMatchSnapshot();
});

test('handles multiple PRs per tag', async () => {
  nock('https://api.github.com')
    .get('/repos/owner/repo/pulls')
    .query({ per_page: 100, state: 'closed' })
    .reply(200, [
      {
        id: 4,
        merged_at: subHours(now, 1),
        number: 4,
        title: 'Pull request 4',
        user: {
          html_url: userHtmlUrl,
          login: userLogin
        }
      },
      {
        id: 3,
        merged_at: subHours(now, 2),
        number: 3,
        title: 'Pull request 3',
        user: {
          html_url: userHtmlUrl,
          login: userLogin
        }
      },
      {
        id: 2,
        merged_at: subHours(now, 4),
        number: 2,
        title: 'Pull request 2',
        user: {
          html_url: userHtmlUrl,
          login: userLogin
        }
      },
      {
        id: 1,
        merged_at: subHours(now, 5),
        number: 1,
        title: 'Pull request 1',
        user: {
          html_url: userHtmlUrl,
          login: userLogin
        }
      }
    ]);

  nock('https://api.github.com')
    .get('/repos/owner/repo/tags')
    .query({ per_page: 100 })
    .reply(200, [{ name: 'v1.0.0', commit: { sha: '1' } }]);

  nock('https://api.github.com')
    .get('/repos/owner/repo/commits')
    .query({ per_page: 100 })
    .reply(200, [
      {
        sha: '1',
        commit: {
          committer: {
            date: subHours(now, 3)
          }
        }
      }
    ]);

  const result = await changelog({
    futureRelease: 'v2.0.0',
    owner,
    repo
  });

  expect(result).toMatchSnapshot();
});

test('can infer repo info', async () => {
  nock('https://api.github.com')
    .get('/repos/owner/repo/pulls')
    .query({ per_page: 100, state: 'closed' })
    .reply(200, [
      {
        id: 1,
        merged_at: subHours(now, 2),
        number: 1,
        title: 'Pull request 1',
        user: {
          html_url: userHtmlUrl,
          login: userLogin
        }
      }
    ]);

  nock('https://api.github.com')
    .get('/repos/owner/repo/tags')
    .query({ per_page: 100 })
    .reply(200, [{ name: 'v1.0.0', commit: { sha: '1' } }]);

  nock('https://api.github.com')
    .get('/repos/owner/repo/commits')
    .query({ per_page: 100 })
    .reply(200, [
      {
        sha: '1',
        commit: {
          committer: {
            date: subHours(now, 1)
          }
        }
      }
    ]);

  const result = await changelog();

  expect(result).toMatchSnapshot();
});

test('throws when unable to get repo info', async () => {
  fs.writeFileSync(
    gitConfigPath,
    `[remote "origin"]\n  url = invalid`,
    'utf-8'
  );

  await expect(changelog()).rejects.toThrow('Unable to parse GitHub url');
});
