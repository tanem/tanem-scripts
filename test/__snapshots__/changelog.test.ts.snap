// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can create a future release 1`] = `
"# Changelog

## [v2.0.0](https://github.com/owner/repo/tree/v2.0.0) (2000-02-01)
[Full Changelog](https://github.com/owner/repo/compare/v1.0.0...v2.0.0)

**Merged pull requests:**

- Pull request 2 [#2](https://github.com/owner/repo/pull/2) ([user](https://github.com/user))

## [v1.0.0](https://github.com/owner/repo/tree/v1.0.0) (2000-01-31)

**Merged pull requests:**

- Pull request 1 [#1](https://github.com/owner/repo/pull/1) ([user](https://github.com/user))
"
`;

exports[`can infer repo info 1`] = `
"# Changelog

## [v1.0.0](https://github.com/owner/repo/tree/v1.0.0) (2000-01-31)

**Merged pull requests:**

- Pull request 1 [#1](https://github.com/owner/repo/pull/1) ([user](https://github.com/user))
"
`;

exports[`handles GitHub API pagination 1`] = `
"# Changelog

## [v2.0.0](https://github.com/owner/repo/tree/v2.0.0) (2000-02-01)
[Full Changelog](https://github.com/owner/repo/compare/v1.0.0...v2.0.0)

**Merged pull requests:**

- Pull request 2 [#2](https://github.com/owner/repo/pull/2) ([user](https://github.com/user))

## [v1.0.0](https://github.com/owner/repo/tree/v1.0.0) (2000-01-31)

**Merged pull requests:**

- Pull request 1 [#1](https://github.com/owner/repo/pull/1) ([user](https://github.com/user))
"
`;

exports[`handles multiple PRs per tag 1`] = `
"# Changelog

## [v2.0.0](https://github.com/owner/repo/tree/v2.0.0) (2000-02-01)
[Full Changelog](https://github.com/owner/repo/compare/v1.0.0...v2.0.0)

**Merged pull requests:**

- Pull request 4 [#4](https://github.com/owner/repo/pull/4) ([user](https://github.com/user))
- Pull request 3 [#3](https://github.com/owner/repo/pull/3) ([user](https://github.com/user))

## [v1.0.0](https://github.com/owner/repo/tree/v1.0.0) (2000-01-31)

**Merged pull requests:**

- Pull request 2 [#2](https://github.com/owner/repo/pull/2) ([user](https://github.com/user))
- Pull request 1 [#1](https://github.com/owner/repo/pull/1) ([user](https://github.com/user))
"
`;

exports[`only uses merged PRs 1`] = `
"# Changelog

## [v2.0.0](https://github.com/owner/repo/tree/v2.0.0) (2000-01-31)
[Full Changelog](https://github.com/owner/repo/compare/v1.0.0...v2.0.0)

**Merged pull requests:**

- Pull request 2 [#2](https://github.com/owner/repo/pull/2) ([user](https://github.com/user))

## [v1.0.0](https://github.com/owner/repo/tree/v1.0.0) (2000-01-30)

**Merged pull requests:**

- Pull request 1 [#1](https://github.com/owner/repo/pull/1) ([user](https://github.com/user))
"
`;

exports[`refetches missing tag commits 1`] = `
"# Changelog

## [v2.0.0](https://github.com/owner/repo/tree/v2.0.0) (2000-02-01)
[Full Changelog](https://github.com/owner/repo/compare/v1.0.0...v2.0.0)

**Merged pull requests:**

- Pull request 1 [#1](https://github.com/owner/repo/pull/1) ([user](https://github.com/user))

## [v1.0.0](https://github.com/owner/repo/tree/v1.0.0) (2000-01-31)

**Merged pull requests:**

- Pull request 2 [#2](https://github.com/owner/repo/pull/2) ([user](https://github.com/user))
"
`;
